test/units/modules/cloud/misc/test_virt_net.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright: (c) 2019, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+import mock
+import pytest
+
+libvirt = pytest.importorskip('libvirt')
```

Comment:
Instead of depending on `libvirt` for the unit tests, have you considered mocking the import?

See https://github.com/ansible/ansible/pull/50850 for an example of how that can be done.

Reply:
Yes, definitely. But I believe this also reduces the value of the unit-test.

Reply:
For testing with actual dependencies such as `libvirt` I recommend using integration tests.

Unit tests can rely on mocking the imports of dependencies, and are good for verifying the internal logic of modules and module_utils. Integration tests use the actual dependencies and are good for verifying they work correctly with modules.

This allows both unit and integration tests to test against stable public APIs, rather than being dependent on the internals of dependencies which are subject to change.

Reply:
Libvirt comes with a test driver ( https://libvirt.org/drvtest.html ) that allow one to test some basic scenarios without doing any actual changes on the system. It does not cover everything but it's already a good start. This is what I use here.

Functional tests are IMO complementary and come with their own limitations (hard to cover all the user-cases, interaction with the local system, extra layers of complexity on top of the code, etc).